src/resource_manager/gateway.py

```diff
@@ -0,0 +1,238 @@
+                    {
+                        "protocol": "HTTP",
+                        "port": 80,
+                        "name": f"{definition.gateway_name}-http-listener",
+                        "hostname": definition.config.external_hostname,
+                        "allowedRoutes": {"namespaces": {"from": "All"}},
+                    },
+                ]
+            },
+        )
+
+        self._set_gateway_class(
+            configured_gateway_class=definition.config.gateway_class, resource=gateway
+        )
+        LOGGER.info("Generated gateway resource: %s", gateway)
+        return gateway
+
+    @_map_k8s_auth_exception
+    def _create_resource(self, resource: GenericNamespacedResource) -> None:
+        """Create a new gateway resource in a given namespace.
+
+        Args:
+            resource: The gateway resource object to create.
+        """
+        self._client.create(resource)
+
+    @_map_k8s_auth_exception
+    def _patch_resource(self, name: str, resource: GenericNamespacedResource) -> None:
+        """Replace an existing gateway resource in a given namespace.
+
+        Args:
+            name: The name of the resource to patch.
+            resource: The modified gateway resource object.
+        """
+        # Patch the resource with server-side apply
+        # force=True is required here so that the charm keeps control of the resource
+        self._client.patch(  # type: ignore[type-var]
+            # mypy can't detect that this is ok for patching custom resources
+            self._gateway_generic_resource,
+            name,
+            resource,
+            patch_type=PatchType.APPLY,
+            force=True,
+        )
+
+    @_map_k8s_auth_exception
+    def _list_resource(self) -> List[GenericNamespacedResource]:
+        """List gateway resources in a given namespace based on a label selector.
+
+        Returns:
+            A list of matched gateway resources.
+        """
+        return list(
+            self._client.list(
+                res=self._gateway_generic_resource, namespace=self._namespace, labels=self._labels
+            )
+        )
+
+    @_map_k8s_auth_exception
+    def _delete_resource(self, name: str) -> None:
+        """Delete a gateway resource from a given namespace.
+
+        Args:
+            name: The name of the V1Ingress resource to delete.
+        """
+        self._client.delete(
+            res=self._gateway_generic_resource, name=name, namespace=self._namespace
+        )
+
+    def gateway_address(self, name: str) -> Optional[str]:
+        """Return the LB address of the gateway resource.
+
+        Poll the address for 100 seconds.
+
+        Args:
+            name (str): name of the gateway resource.
+
+        Returns:
+            Optional[str]: The addresses assigned to the gateway object, or none.
+        """
+        deadline = time.time() + 60
+        delay = 5
+        gateway_address = None
+        while time.time() < deadline:
```

Comment:
Shouldnt we use some kind of library to do that retry flow?
@gregory-schiano 
Postgresql is using tenacity
https://github.com/canonical/postgresql-k8s-operator/blob/04c217e70b2e41d5a68fdc604404edc3dcb0d5b7/src/upgrade.py#L107

Reply:
I created a ticket for this (ISD-2002 - internal). Will make a follow up PR for it as I agree that a library would work best for this. I'm closing this conversation and we can move our discussion internally if needed.